File: frontend/src/types/budget.ts
export type BudgetStatus = "healthy" | "warning" | "near_limit" | "over_budget";

// Matches the backend's BudgetOut schema exactly - usage figures are
// computed server-side at read time, never stored.
export interface Budget {
  id: number;
  category_id: number;
  category_name: string;
  category_icon: string | null;
  amount: number;
  month: number; // 1-12
  year: number;
  spent: number;
  remaining: number;
  percent_used: number;
  status: BudgetStatus;
  alert_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface BudgetInput {
  category_id: number;
  amount: number;
  month: number;
  year: number;
}

File: backend/app/schemas/budget.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

BudgetStatus = Literal["healthy", "warning", "near_limit", "over_budget"]


class BudgetBase(BaseModel):
    category_id: int
    amount: float = Field(..., gt=0, description="Amount must be greater than zero")
    month: int = Field(..., ge=1, le=12)
    year: int = Field(..., ge=2000, le=2100)


class BudgetCreate(BudgetBase):
    pass


class BudgetUpdate(BudgetBase):
    pass


class BudgetOut(BudgetBase):
    """
    Includes usage figures computed at read time from the category's
    expense transactions for that month/year - never stored, so it's
    always current (PRD sections 19-21).
    """

    model_config = ConfigDict(from_attributes=True)

    id: int
    category_name: str
    category_icon: str | None
    spent: float
    remaining: float
    percent_used: float
    status: BudgetStatus
    alert_message: str | None
    created_at: datetime
    updated_at: datetime
